File: src/dependent_types/bruijn.hs
{-
  An implementation of substitution, normalization via beta-reduction, and a
  simple implementation of type inference (NOT normalization; in other words,
  all lambdas are type-annotated) for dependent types.

  The code here is largely adapated from the following article:
    http://math.andrej.com/2012/11/08/how-to-implement-dependent-type-theory-i/
  except retrofitted to work with de bruijn indices, and a (possibly naiive?)
  implementation of cumulative universes.
-}

{-# LANGUAGE OverloadedStrings, DeriveFunctor #-}

import qualified Data.Text as T
import qualified Data.Map as M
import Control.Monad.Except
import Control.Monad.State

-- Helper Defintions

tshow :: Show a => a -> T.Text
tshow = T.pack . show

-- Core Syntax

data AbsT a
  = Abs T.Text (ExpT a) (ExpT a)
  deriving (Eq, Functor)
type Abs = AbsT Int

data ExpT a 
  = Ref T.Text
  | Idx a
  | Lam (AbsT a)
  | App (ExpT a) (ExpT a)
  | Unv Int
  | Pi  (AbsT a)
  | Tup (AbsT a) (ExpT a)
  | Sgm (AbsT a)
  deriving (Eq, Functor)
type Exp = ExpT Int

instance Show a => Show (AbsT a) where
  show (Abs x (Lam a) e) = "(" ++ show (Lam a) ++ "). " ++ show e
  show (Abs x (Pi a) e) = "(" ++ show (Pi a) ++ "). " ++ show e
  show (Abs x t e) = show t ++ ". " ++ show e

instance Show a => Show (ExpT a) where
  show (Ref n) = T.unpack n
  show (Idx i) = show i
  show (Lam a) = "λ" ++ show a
  show (App (Lam a) b) = "(" ++ show (Lam a) ++ ") " ++ show b
  show (App a (App b c)) = show a ++ " (" ++ show (App b c) ++ ")"
  show (App a b) = show a ++ " " ++ show b
  show (Pi  a) = "Π" ++ show a
  show (Unv u) = "Set" ++ show u
  show (Tup (Abs _ t b) a) = "(" ++ show a ++ ": " ++ show t ++ ", " ++ show b ++ ")"
  show (Sgm a) = "Σ" ++ show a

-- Substitution

substRaw :: Int -> Exp -> Exp -> Exp
substRaw k w (Ref n) = Ref n
substRaw k w (Idx i)
  | i == k    = w
  | i > k     = Idx (i - 1)
  | otherwise = Idx i
substRaw k w (Lam (Abs x t e)) =
  Lam (Abs x (substRaw k w t) (substRaw (k + 1) w e))
substRaw k w (App a b) =
  App (substRaw k w a) (substRaw k w b)
substRaw k w (Pi (Abs x t e)) =
  Pi (Abs x (substRaw k w t) (substRaw (k + 1) w e))
substRaw k w (Unv u) = Unv u
substRaw k w (Tup (Abs x t b) a) =
  Tup (Abs x (substRaw k w t) (substRaw (k + 1) w t)) (substRaw k w a)
substRaw k w (Sgm (Abs x t e)) =
  Sgm (Abs x (substRaw k w t) (substRaw (k + 1) w e))

subst :: Exp -> Exp -> Exp
subst = substRaw 0

-- Normalization

type Gam = ([Exp], M.Map T.Text (Exp, Maybe Exp))

gempty :: Gam
gempty = ([], M.empty)

norm :: Gam -> Exp -> Either T.Text Exp
norm (_, m) (Ref n) =
  case M.lookup n m of
    Just (_, Just v) -> Right v
    Just (_, Nothing) -> Right $ Ref n
    Nothing -> Left $ "Undefined identifier: " <> n
norm g (Idx i) = Right $ Idx i
norm g (Lam (Abs x t e)) = do
  t' <- norm g t
  e' <- norm g e
  Right . Lam $ Abs x t' e'
norm g (App a b) = do
  a' <- norm g a
  b' <- norm g b
  case a' of
    Lam (Abs x t e) -> norm g $ subst b' e
    _ -> Right $ App a' b'
norm g (Pi (Abs x t e)) = do
  t' <- norm g t
  e' <- norm g e
  Right . Pi $ Abs x t' e'
norm g (Unv u) = Right $ Unv u
norm g (Tup (Abs x t b) a) = do
  t' <- norm g t
  a' <- norm g a
  b' <- norm g b
  Right $ Tup (Abs x t' b') a'
norm g (Sgm (Abs x t e)) = do
  t' <- norm g t
  e' <- norm g e
  Right . Sgm $ Abs x t' e'

-- Type Passing

passes :: Exp -> Exp -> Bool
passes (Unv a) (Unv b) = a <= b
passes a b = a == b

checkPasses :: Exp -> Exp -> Either T.Text ()
checkPasses a b =
  if passes a b then
    Right ()
  else
    Left $ "Attempt to pass `" <> tshow a <> "` as `" <> tshow b <> "`."

-- Inference

inferUnv :: Gam -> Exp -> Either T.Text Int
inferUnv g e = do
  t <- infer g e
  t' <- norm g t
  case t' of
    Unv k -> Right k
    _ -> Left "Type expected."

inferPi :: Gam -> Exp -> Either T.Text Abs
inferPi g e = do
  t <- infer g e
  t' <- norm g t
  case t' of
    Pi a -> Right a
    _ -> Left "Function expected."

infer :: Gam -> Exp -> Either T.Text Exp
infer (_, m) (Ref n) =
  case M.lookup n m of
    Just (t, _) -> Right t
    Nothing -> Left $ "Undefined identifier: " <> n
infer (s, _) (Idx i) = 
  if i < length s then
    Right . fmap (+ succ i) $ (s !! i)
  else
    Left $ "Index out of bounds: " <> tshow i
infer g@(s, m) (Lam (Abs x t e)) = do
  t' <- norm g t
  _ <- inferUnv g t'
  et <- infer (t' : s, m) e
  Right . Pi $ Abs x t' et
infer g@(s, m) (App a b) = do
  a' <- norm g a
  Abs x t e <- inferPi g a'
  b' <- norm g b
  bt <- infer g b'
  checkPasses bt t
  Right $ subst b' e
infer g@(s, m) (Pi (Abs x t e)) = do
  t' <- norm g t
  tu <- inferUnv g t'
  eu <- inferUnv (t' : s, m) e
  Right . Unv $ max tu eu
infer g (Unv u) = Right $ Unv (u + 1)
infer g@(s, m) (Tup (Abs x t b) a) = do
  t' <- norm g t
  _ <- inferUnv g t'
  at <- infer g a
  checkPasses at t'
  bt <- infer (t' : s, m) b
  Right . Sgm $ Abs x t' bt
infer g@(s, m) (Sgm (Abs x t e)) = do
  t' <- norm g t
  e' <- norm g e
  tu <- inferUnv g t'
  eu <- inferUnv g e'
  Right . Unv $ max tu eu

-- Examples

lam :: T.Text -> Exp -> Exp -> Exp
lam x t e = Lam $ Abs x t e

pit :: T.Text -> Exp -> Exp -> Exp
pit x t e = Pi  $ Abs x t e

-- (x : T = a, b)
tup :: T.Text -> Exp -> Exp -> Exp -> Exp
tup x t a b = Tup (Abs x t b) a

eid :: Exp
eid = lam "t" (Unv 1) . lam "x" (Idx 0) $ (Idx 0)

edollar :: Exp
edollar = lam "t" (Unv 1) 
        . lam "f" (pit "y" (Idx 0) (Idx 1))
        . lam "x" (Idx 1) 
        $ App (Idx 1) (Idx 0)

exy :: Exp
exy = lam "t" (Unv 1)
    . lam "x" (Idx 0)
    . lam "y" (Idx 1)
    $ tup "x" (Idx 2) (Idx 1) (Idx 1) 

-- Main Program

test :: String -> Gam -> Exp -> IO ()
test n g e = do
  putStrLn $ "-- " ++ n ++ " --"
  putStrLn $ "Raw:        " ++ show e
  putStrLn $ "Norm: " ++ (show $ norm g e)
  putStrLn $ "Type: " ++ (show $ infer g e)

main = do
  test "id" gempty eid
  test "id Unv 0" gempty (App eid (Unv 0))
  test "$" gempty edollar
  test "$ Unv 0" gempty (App edollar (Unv 0))
  test "xy" gempty exy
  test "xy Unv 0" gempty (App exy (Unv 0))

File: src/dependent_types/bruijn.ml
(*
** An implementation of substitution, normalization via beta-reduction, and a
** simple implementation of type inference (NOT elaboration; in other words,
** all lambdas are type-annotated) for dependent types.
**
** The code here is largely adapated from the following article:
**   http://math.andrej.com/2012/11/08/how-to-implement-dependent-type-theory-i/
** except retrofitted to work with de bruijn indices, and a (possibly naiive?)
** implementation of cumulative universes.
**
** This version is written in Amulet: https://amulet.works/
*)

include import "prelude.ml"

(*
** Core
*)

type nam <- string
type exp_t 't =                                      (* D ::=            *)
  | Ref of nam                                       (* | x              *)
  | Idx of 't                                        (* | #              *)
  | Lam of nam      * exp_t 't * exp_t 't            (* | λx: D. D       *)
  | App of exp_t 't * exp_t 't                       (* | D D            *)
  | Pi  of nam      * exp_t 't * exp_t 't            (* | Πx: D. D       *)
  | Unv of int                                       (* | U0, U1, U2, ...*)
  | Tup of nam      * exp_t 't * exp_t 't * exp_t 't (* | (x : D = D, D) *)
  | Sgm of nam      * exp_t 't * exp_t 't            (* | Σx: D. D       *)
type exp <- exp_t int

(*
** Typeclass Instances
*)

instance eq 't => eq (exp_t 't) begin
  let a == b =
    match a, b with
    | Ref a, Ref b -> a == b
    | Idx a, Idx b -> a == b
    | Lam (an, at, ae), Lam (bn, bt, be) -> an == bn && at == bt && ae == be
    | App (a, b), App (c, d) -> a == c && b == d
    | Pi (an, at, ae), Pi (bn, bt, be) -> an == bn && at == bt && ae == be
    | Unv a, Unv b -> a == b
    | Tup (an, at, aa, ab), Tup (bn, bt, ba, bb) ->
      an == bn && at == bt && aa == ba && ab == bb
    | Sgm (an, at, ae), Sgm (bn, bt, be) -> an == bn && at == bt && ae == be
    | _, _ -> false
end

instance show 'a => show (exp_t 'a) begin
  let show = function
    | Ref n         -> n
    | Idx i         -> show i
    | Lam (_, t, e) -> "\\" ^ show t ^ "]. " ^ show e
    | App (a, b)    -> "(" ^ show a ^ ") " ^ show b
    | Pi (_, t, e)  -> "^" ^ show t ^ ". " ^ show e
    | Unv u         -> "Set" ^ show u
    | Tup (_, t, a, b) -> "(" ^ show a ^ ": " ^ show t ^ ", " ^ show b ^ ")"
    | Sgm (_, t, e) -> "&" ^ show t ^ ". " ^ show e
end

instance functor exp_t begin
  let f <$> e =
    match e with
    | Ref n            -> Ref n
    | Idx i            -> Idx (f i)
    | Lam (n, t, e)    -> Lam (n, f <$> t, f <$> e)
    | App (a, b)       -> App (f <$> a, f <$> b)
    | Pi (n, t, e)     -> Pi  (n, f <$> t, f <$> e)
    | Unv u            -> Unv u
    | Tup (n, t, a, b) -> Tup (n, f <$> t, f <$> a, f <$> b)
    | Sgm (n, t, e)    -> Sgm (n, f <$> t, f <$> e)
end

(*
** Substitution
*)

let subst_var k w i =
  if i == k then
    w
  else if i > k then
    Idx (i - 1)
  else
    Idx i

let rec subst_raw k w = function
  | Ref n            -> Ref n
  | Idx i            -> subst_var k w i
  | Lam (n, t, e)    -> Lam (n, subst_raw k w t, subst_raw (k + 1) w e)
  | App (a, b)       -> App (subst_raw k w a, subst_raw k w b)
  | Pi (n, t, e)     -> Pi  (n, subst_raw k w t, subst_raw (k + 1) w e)
  | Unv u            -> Unv u
  | Tup (n, t, a, b) ->
    Tup (n, subst_raw k w t, subst_raw k w a, subst_raw (k + 1) w b)
  | Sgm (n, t, e)    -> Sgm (n, subst_raw k w t, subst_raw (k + 1) w e)
let subst = subst_raw 0

(*
** Normalization
*)

type gam <- (list exp, list (nam * (exp * option exp)))
let gempty = ([], [])

let rec norm (s, m) = function
  | Ref n ->
    match lookup n m with
    | Some (_, Some x) -> Right x
    | Some (_, None)   -> Right @@ Ref n
    | None -> Left @@ "Unknown identifier: " ^ n
  | Idx i -> Right @@ Idx i
  | Lam (n, t, e) -> begin
      with t' <- norm (s, m) t
      with e' <- norm (s, m) e
      Right @@ Lam (n, t', e')
    end
  | App (a, b) -> begin
      with a' <- norm (s, m) a
      with b' <- norm (s, m) b
      Right @@ App (a', b')
    end
  | Pi (n, t, e) -> begin
      with t' <- norm (s, m) t
      with e' <- norm (s, m) e
      Right @@ Pi (n, t', e')
    end
  | Unv u -> Right @@ Unv u
  | Tup (n, t, a, b) -> begin
      with t' <- norm (s, m) t
      with a' <- norm (s, m) a
      with b' <- norm (s, m) b
      Right @@ Tup (n, t', a', b')
    end
  | Sgm (n, t, e) -> begin
      with t' <- norm (s, m) t
      with e' <- norm (s, m) e
      Right @@ Sgm (n, t', e')
    end

(*
** Type Inference + Checking
*)

let passes a b =
  match a, b with
  | Unv a, Unv b -> a <= b
  | a, b -> a == b

let check_passes a b =
  if passes a b then
    Right ()
  else
    Left @@ "Attempt to pass " ^ show a ^ " as " ^ show b

let rec infer (s, m) = function
  | Ref n ->
    match lookup n m with
    | Some (t, _) -> Right t
    | None -> Left @@ "Unknown identifier: " ^ n
  | Idx i ->
    if i < length s then
      Right @@ (+ (i + 1)) <$> nth i s
    else
      Left @@ "Index out of bounds: " ^ show i
  | Lam (n, t, e) -> begin
      with t' <- norm (s, m) t
      with _ <- infer_unv (s, m) t'
      with et <- infer (t' :: s, m) e
      Right @@ Pi (n, t', et) 
    end
  | App (a, b) -> begin
      with a' <- norm (s, m) a
      with (_, t, e) <- infer_pi (s, m) a'
      with b' <- norm (s, m) b
      with bt <- infer (s, m) b'
      with _ <- check_passes bt t
      Right @@ subst b' e
    end
  | Pi (_, t, e) -> begin
      with t' <- norm (s, m) t
      with tu <- infer_unv (s, m) t'
      with eu <- infer_unv (t' :: s, m) e
      Right @@ Unv @@ (if tu > eu then tu else eu)
    end
  | Unv u -> Right @@ Unv (u + 1)
  | Tup (x, t, a, b) -> begin
      with t' <- norm (s, m) t
      with _ <- infer_unv (s, m) t'
      with at <- infer (s, m) a
      with _ <- check_passes at t'
      with bt <- infer (s, m) b
      Right @@ Sgm (x, t', bt)
    end
  | Sgm (_, t, e) -> begin
      with t' <- norm (s, m) t
      with e' <- norm (s, m) e
      with tu <- infer_unv (s, m) t'
      with eu <- infer_unv (s, m) e'
      Right @@ Unv @@ (if tu > eu then tu else eu)
    end
and infer_unv (s, m) e = begin
    with t <- infer (s, m) e
    with t' <- norm (s, m) t
    match t' with
    | Unv u -> Right u
    | _ -> Left "Expected type."
  end
and infer_pi (s, m) e = begin
    with t <- infer (s, m) e
    with t' <- norm (s, m) t
    match t' with
    | Pi (n, t, e) -> Right (n, t, e)
    | _ -> Left "Expected function."
  end


let lam x t e = Lam (x, t, e)
let pit x t e = Pi (x, t, e)
let app a b = App (a, b)

let eid =
  lam "t" (Unv 1) (
    lam "x" (Idx 0) (
      Idx 0
    )
  )

let edollar =
  lam "t" (Unv 1) (
    lam "f" (pit "ft" (Idx 0) (Idx 1)) (
      lam "x" (Idx 1) (
        app (Idx 1) (Idx 0)
      )
    )
  )

let test n e =
  put_line @@ " -- " ^ n ^ " --"
  put_line @@ "Raw:  " ^ show e
  put_line @@ "Norm: " ^ (show @@ from_right e @@ infer gempty e)

let () =
  test "$" edollar
  test "$ Unv 0" (app edollar (Unv 0))
  test "id" eid
  test "id Unv 0" (app eid (Unv 0))
